### PR TITLE
tailscale: fix tailscale ssh

### LINKED ISSUE #310950: tailscale - combining binaries breaks `tailscale ssh`
### Describe the bug

Commit "tailscale: combine client and daemon binary" breaks `tailscale ssh`

* https://github.com/NixOS/nixpkgs/commit/ccf4d0c623495168171bee72f511d7e56b4cdc84

```console
╭─    ~/nix    main *1 +3                                                                                                                   ✔  vesu@glacier  18:59:02  
╰─ tailscale ssh myhost
2024/05/11 18:59:06 tailscaled does not take non-flag arguments: ["nc" "myhost.word-other.ts.net." "22"]
Connection closed by UNKNOWN port 65535
╭─    ~/nix    main *1 +3                                                                                                               255 ✘  vesu@glacier  18:59:06  
╰─ git au && just switch
sudo nixos-rebuild switch --flake ".#glacier"
warning: Git tree '/home/vesu/nix' is dirty
warning: updating lock file '/home/vesu/nix/flake.lock':
• Updated input 'nixpkgs-test':
    'github:nixos/nixpkgs/ccf4d0c623495168171bee72f511d7e56b4cdc84' (2024-04-24)
  → 'github:nixos/nixpkgs/cd7bb6126ced1216fb49f48103db7811bd308a39' (2024-04-22)
warning: Git tree '/home/vesu/nix' is dirty
building the system configuration...
warning: Git tree '/home/vesu/nix' is dirty
activating the configuration...
setting up /etc...
reloading user units for vesu...
setting up tmpfiles
reloading the following units: dbus.service
restarting the following units: tailscaled.service
╭─    ~/nix    main *1 +3 !1                                                                                                        ✔  32s   vesu@glacier  19:00:11  
╰─ tailscale version    
1.64.2
  go version: go1.22.2
╭─    ~/nix    main *1 +3 !1                                                                                                                ✔  vesu@glacier  19:00:16  
╰─ tailscale ssh myhost 
[vesu@<ip> ~]$ 
logout
Connection to myhost.word-other.ts.net. closed.

```

### Steps To Reproduce

Attempt to use `tailscale ssh ...` with nixpkgs commit `ccf4d0c623495168171bee72f511d7e56b4cdc84` or later and notice that it fails with the error ` tailscaled does not take non-flag arguments: ["nc" "myhost.word-other.ts.net." "22"]`

### Expected behavior

`tailscale ssh ...` should not fail, as in commits prior.

### Notify maintainers

@SuperSandro2000 

### Metadata

```console
╰─ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.87, NixOS, 23.11 (Tapir), 23.11.20240511.71bae31`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.05"`
 - channels(vesu): `"home-manager-23.05.tar.gz"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by SuperSandro2000**
Closes #310950

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by endocrimes** [comment]
We should revert the combination PR rather than shipping an un-reviewed PR as a patch - We clearly don't have enough testing infrastructure in place here, and repeatedly breaking folks networking and security boundary software is painfully invasive.

**Comment by endocrimes** [comment]
(Reviewed the patch and there's technically also a slight regression).

File: pkgs/servers/tailscale/default.nix
Comment (mfrw) [inline pkgs/servers/tailscale/default.nix:31]: I was wondering if we should wait for the review in the upstream and then include this patch :)

Comment (SuperSandro2000) [reply pkgs/servers/tailscale/default.nix:31]: This is a none security relevant change and we understand the underlying issue pretty well, so we don't necessarily need the upstream review.

**Comment by SuperSandro2000** [comment]
> (Reviewed the patch and there's technically also a slight regression).

That was easy to fix by using the same method the tailscaled uses to detect if it should start the cli https://github.com/tailscale/tailscale/pull/5193/files

> We should revert the combination PR rather than shipping an un-reviewed PR as a patch - We clearly don't have enough testing infrastructure in place here, and repeatedly breaking folks networking and security boundary software is painfully invasive.

"We" as in tailscale. I took this from their upstream documentation https://tailscale.com/kb/1207/small-tailscale and since there were no hints at all, I assumed this is straightforward and easy to use to reduce unnecessary binary size.

I think we understand the problem at hand pretty well and I have looked at all other usages of os.Executable() and there should not be affected by this, so a revert would create a noisy history.

**Comment by endocrimes** [comment]
We: Nixpkgs - There have been several ways that we've broken the tailscale package and module lately - including this. We should move back to a stable point and then only start making changes _if they're going to be well tested first_.

**Comment by SuperSandro2000** [comment]
Well, the other breakage was because a years old commit got reverted and that PR shouldn't have been merged in the first place and didn't receive a proper review from a maintainer.

We cannot realistically test everything, hence we always need to rely a bit on nixos-unstable to discover some issues.

I think we should just merge this and move forward and then we are at the stable point again.

**Review by 06kellyjac — APPROVED** [review]

**Comment by 06kellyjac** [comment]
I agree that not every tailscale feature can be realistically tested every PR.
It could be argued that tailscale ssh should have been tested but it can also be equally argued that nixos-unstable has no formal stability guarantees and by having any integration testing via headscale is ahead of other distributions.

If this issue when using `ts_include_cli`, and symlinking as described in the official tailscale docs, is not exclusive to the nix built copy due to our wrapping, this request for more extensive testing could also be applied to tailscale rather than unpaid volunteers.

---

In order to focus on how we (as the nixpkgs maintainers for tailscale) can improve going forwards rather than focusing on past negatives here are some proposals:

* we could look to extend headscale integration tests to further tailscale features
* we could leave non-version-bump changes open longer for people who are interested in testing
  * duration to be determined, could have exceptions
  * I still think faster merging of updates should be fine, extension of headscale integration testing should assist, we could pick a duration but it must be quicker than non-version-bump changes IMO
* people concerned with the stability of tailscale could join in testing these PRs and post their experiences
  * PRs being open longer allows for more opertunity
  * individuals could be informal testers or officially sign up as maintainers
  * a method of notifying these individuals (including informal testers) would potentially be necessary

We will not catch all issues but these could help. I welcome thoughts on these proposals or additional ideas.

**Pull request merged**
